File: languages/dotnet/tests/Revaly.Sdk.Tests/NoRetryTests.cs
using Revaly.Sdk.Errors;
using Revaly.Sdk.Testing;
using Revaly.Sdk.Tests.TestSupport;

namespace Revaly.Sdk.Tests;

/// <summary>
/// ADR-SDK-004: no hidden retries anywhere — single-shot semantics except the explicit,
/// caller-bounded reconcile loop. One call = exactly one wire request, on every path.
/// </summary>
public class NoRetryTests
{
    [Fact]
    public async Task A_successful_charge_sends_exactly_one_request()
    {
        var mock = new RapMockTransport();
        mock.Charge().ReturnsApproved();
        using var client = TestClient.Create(mock);

        await client.Payments.ChargePaymentAsync(TestClient.ChargeRequest());

        Assert.Single(mock.Requests);
    }

    [Fact]
    public async Task An_OutcomeUnknown_500_is_never_retried()
    {
        var mock = new RapMockTransport();
        mock.Charge().Returns500();
        using var client = TestClient.Create(mock);

        await Assert.ThrowsAsync<OutcomeUnknownException>(
            () => client.Payments.ChargePaymentAsync(TestClient.ChargeRequest()));

        Assert.Single(mock.Requests);
    }

    [Fact]
    public async Task A_TransientFailure_is_never_retried_failover_is_the_callers_move()
    {
        var mock = new RapMockTransport();
        mock.Charge().ReturnsNotProcessed503();
        using var client = TestClient.Create(mock);

        await Assert.ThrowsAsync<TransientFailureException>(
            () => client.Payments.ChargePaymentAsync(TestClient.ChargeRequest()));

        Assert.Single(mock.Requests);
    }

    [Fact]
    public async Task Reconcile_with_a_single_attempt_policy_sends_exactly_one_request()
    {
        var mock = new RapMockTransport();
        mock.Reconcile(SyntheticData.MerchantTransactionId).ReturnsNotFoundYet();
        using var client = TestClient.Create(mock);

        await client.ReconcileAsync(
            SyntheticData.MerchantTransactionId,
            new Reconcile.ReconcilePolicy(1, TimeSpan.FromSeconds(5), TimeSpan.Zero));

        Assert.Single(mock.Requests);
    }
}

File: languages/java/tests/src/test/java/co/revaly/sdk/NoRetryTests.java
package co.revaly.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import co.revaly.sdk.errors.OutcomeUnknownException;
import co.revaly.sdk.errors.RapCoreException;
import co.revaly.sdk.reconcile.ReconcilePolicy;
import co.revaly.sdk.testing.RapMockTransport;
import java.time.Duration;
import org.junit.jupiter.api.Test;

/**
 * ADR-SDK-004 / failover-contract §5: no hidden retries anywhere, single-shot semantics — the
 * explicit caller-bounded reconcile re-poll is the ONLY loop this SDK owns.
 */
class NoRetryTests {

    @Test
    void aFailedChargeSendsExactlyOneRequest() {
        RapMockTransport mock = new RapMockTransport();
        mock.charge().returnsServerError();
        RapClient client = TestClient.client(mock);

        assertThrows(RapCoreException.class, () -> client.charge(TestClient.paymentRequest()));

        assertEquals(1, mock.getRequests().size());
    }

    @Test
    void aTransientFailureIsNotRetriedTheCallerFailsOver() {
        RapMockTransport mock = new RapMockTransport();
        mock.charge().returnsNotProcessed503();
        RapClient client = TestClient.client(mock);

        assertThrows(RapCoreException.class, () -> client.charge(TestClient.paymentRequest()));

        assertEquals(1, mock.getRequests().size());
    }

    @Test
    void aDeadlineExpiryIsNotRetriedResubmissionIsForbidden() {
        RapMockTransport mock = new RapMockTransport();
        mock.charge().throwsTimeoutAfterSend();
        RapClient client = TestClient.client(mock);

        assertThrows(
                OutcomeUnknownException.class, () -> client.charge(TestClient.paymentRequest()));

        assertEquals(
                1, mock.getRequests().size(), "resubmitting an unknown outcome can double-charge");
    }

    @Test
    void reconcileIsCallerBoundedAndStopsAtMaxAttempts() throws Exception {
        RapMockTransport mock = new RapMockTransport();
        mock.reconcile(TestClient.MTX).notFoundYet();
        RapClient client = TestClient.client(mock);

        client.reconcile(
                TestClient.MTX,
                new ReconcilePolicy(3, Duration.ofSeconds(10), Duration.ofMillis(1)));

        assertEquals(3, mock.getRequests().size());
    }
}
